curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md

```diff
@@ -0,0 +1,128 @@
+---
+id: 5eb257ba998e23057d209754
+title: Run-length Encoding
+challengeType: 5
+---
+
+## Description
+<section id='description'>
+Given a string containing uppercase characters (A-Z), compress repeated 'runs' of the same character by storing the length of that run, and provide a function to reverse the compression.
+</section>
+
+## Instructions
+<section id='instructions'>
+Write two functions <code>encode</code> and <code>decode</code>.
+
+<code>encode(input)</code> takes a string and return the run-length encoding as an array
+containing pairs with the first element representing the number of occurences and the second
+element representing the letter.
+
+<code>decode(input)</code> takes an array of pairs representing a run-length encoded string
+and returns the decoded string.
+
+You may assume that the input will only contain uppercase letters.
+
+Example:
+<br>
+Original: <code>ABBBCDDDD</code>
+<br>
+Encoded: <code>[[1,'A'], [3,'B'], [1,'C'], [4,'D']]</code>
+
+</section>
+
+## Tests
+<section id='tests'>
+
+``` yml
+tests:
+  - text: <code>encode</code> should return an array.
+    testString: assert(typeof encode(input_1) == 'object');
+  - text: <code>decode</code> should return a string.
+    testString: assert(typeof decode(output_1) == 'string');
+  - text: <code>encode</code> should work for <code>Test 1</code>
+    testString: assert(arrayEquals(encode(input_1), output_1))
+  - text: <code>decode</code> should work for <code>Test 1</code>
+    testString: assert(decode(output_1) === input_1)
+  - text: <code>encode</code> should work for <code>Test 2</code>
+    testString: assert(arrayEquals(encode(input_2), output_2))
+  - text: <code>decode</code> should work for <code>Test 2</code>
+    testString: assert(decode(output_2) === input_2)
+  - text: <code>encode</code> should work for <code>Test 3</code>
+    testString: assert(arrayEquals(encode(input_3), output_3))
+  - text: <code>decode</code> should work for <code>Test 3</code>
+    testString: assert(decode(output_3) === input_3)
```

Comment:
Using the Before Test section to simplify the `testString`'s here is a great idea. But we should make each test case clear in the `text` for each of the tests:

````
tests:
  - text: <code>encode</code> should return an array.
    testString: assert(typeof encode(input_1) === 'object');
  - text: <code>decode</code> should return a string.
    testString: assert(typeof decode(output_1) === 'string');
  - text: <code>encode("ABBBCDDDD")</code> should return <code>[[1, 'A'], [3, 'B'], [1, 'C'], [4, 'D']]</code>
    testString: assert(arrayEquals(encode(input_1), output_1))
  - text: <code>decode([[1, 'A'], [3, 'B'], [1, 'C'], [4,'D']])</code> should return <code>"ABBBCDDDD"</code>
    testString: assert(decode(output_1) === input_1)
  - text: <code>encode("WWWWWWWWWWWWBWWWWWWWWWWWWBBBWWWWWWWWWWWWWWWWWWWWWWWWBWWWWWWWWWWWWWW")</code> should return <code>[[12, 'W'], [1, 'B'], [12, 'W'], [3, 'B'], [24, 'W'], [1, 'B'], [14, 'W']]</code>
    testString: assert(arrayEquals(encode(input_2), output_2))
  - text: <code>decode([[12, 'W'], [1, 'B'], [12, 'W'], [3, 'B'], [24, 'W'], [1, 'B'], [14, 'W']])</code> should return <code>"WWWWWWWWWWWWBWWWWWWWWWWWWBBBWWWWWWWWWWWWWWWWWWWWWWWWBWWWWWWWWWWWWWW"</code>
    testString: assert(decode(output_2) === input_2)
  - text: <code>encode("")</code> should return <code>[]</code>
    testString: assert(arrayEquals(encode(input_3), output_3))
  - text: <code>decode([])</code> should return <code>""</code>
    testString: assert(decode(output_3) === input_3)
````